scripts/train_lm.py

```diff
@@ -144,15 +144,70 @@ def WriteMetaparameters(metaparameters, ngram_order, num_train_sets, out_file):
             i += 1
     f.close()
 
-def FormatMetaparameters(metaparameters):
+def FormatMetaparameters(metaparameters, num_train_sets):
+    # store metaparameters as floats during checking and rerounding
     out = []
+    # final output metaparameters in string
+    out_param = []
+    # round all float metaparameters to 3rd decimal places 
+    round_index = 3
     for param in metaparameters:
-        x = '{:.3f}'.format(param)
-        if x == '0.00':
-            x = '{:.3g}'.format(param)
+        x = round(param, round_index)
         out.append(x)
-
-    return ','.join(out)
+    
+    size = len(metaparameters)
+    
+    # check1: whether 0.000 or 1.000 occur because of rounding
+    # after this step, no 0.000 or 1.000 occurs unless the original number equal
+    # to 0 or 1 (if this happen, the parameters are invalid and it exits 
+    # with a warning, while this case can be very rare)
+    for i in range(0, size): 
+        if out[i] in [0.0, 1.0]:
+            out[i] = RoundToProperDigit(metaparameters[i], round_index)
+            if out[i] in [0.0, 1.0]:
+                sys.exit("train_lm.py: there is a {0}. metaparameters should be in range (0, 1).".format(out[i]))
+    
+    # check2: check repeating values. if happen, then reround them until they are different
+    # if there exist same values originally, exit with a warning (this can be very rare) 
+    # after this step, there should be no any repeating values in set out
+    for i in range(0, size):
+        # Note: the first #num_train_sets weights are not needed to be checked (they can be the same)
+        if i < num_train_sets:
+            out_param.append(format(out[i]))
+        else:
+            x = out[i]
+            for j in range(i + 1, size):
+                y = out[j]
+                # we will round to the next digit when we detect repeated values
+                round_digit = round_index + 1
+                round_max_digit = max(len(str(metaparameters[i])) - 2, len(str(metaparameters[j])) -2)
```

Comment:
Ok. I should have thought about it. But I found the script
validate_metaparameters.py has done this work (1) range in (0,1); 2)
relationship of data-type weights etc) . If users provide some wired
numbers, they can be detected by this script. So I guess validating them in
ParseMetaparameters() is redundant. How do you think?

On Thu, Aug 18, 2016 at 11:32 PM, Wang Jian <notifications@github.com>
wrote:

> In scripts/train_lm.py
> <https://github.com/danpovey/pocolm/pull/52#discussion_r75423580>:
>
> > +                sys.exit("train_lm.py: there is a {0}. metaparameters should be in range (0, 1).".format(out[i]))
> > +
> > +    # check2: check repeating values. if happen, then reround them until they are different
> > +    # if there exist same values originally, exit with a warning (this can be very rare)
> > +    # after this step, there should be no any repeating values in set out
> > +    for i in range(0, size):
> > +        # Note: the first #num_train_sets weights are not needed to be checked (they can be the same)
> > +        if i < num_train_sets:
> > +            out_param.append(format(out[i]))
> > +        else:
> > +            x = out[i]
> > +            for j in range(i + 1, size):
> > +                y = out[j]
> > +                # we will round to the next digit when we detect repeated values
> > +                round_digit = round_index + 1
> > +                round_max_digit = max(len(str(metaparameters[i])) - 2, len(str(metaparameters[j])) -2)
>
> Oh yes, it may be not necessary to check the input for
> FormatMetaparameters(). But it would be better if we check the args in
> ParseMetaparameters(), because people can set --bypass-metaparameter-
> optimization to any possible value, which is not controlled by us.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/danpovey/pocolm/pull/52/files/87ca2779cd82ad774f490c940435f54fe5c991b7#r75423580>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ANVxSoSRq24M1FWdCoOin1YJa5MavXMQks5qhSPRgaJpZM4Jl6Bt>
> .
>



-- 
Ke Li
Dept. of Electrical and Computer Engineering
Johns Hopkins University
Email: kli26@jhu.edu